internal/providercache/installer_events.go

```diff
@@ -0,0 +1,125 @@
+package providercache
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform/addrs"
+	"github.com/hashicorp/terraform/internal/getproviders"
+)
+
+// InstallerEvents is a collection of function references that can be
+// associated with an Installer object in order to be notified about various
+// installation lifecycle events during an install operation.
```

Comment:
This is a really nice pattern that I hadn't seen before 👏

Reply:
I think the place I saw it first is [`httptrace.ClientTrace`](https://godoc.org/net/http/httptrace#ClientTrace).